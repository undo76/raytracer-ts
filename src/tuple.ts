export type tuple = number[];

export const vector = (x: number, y: number, z: number): tuple => [x, y, z, 0];
export const point = (x: number, y: number, z: number): tuple => [x, y, z, 1];

export type color = [number, number, number];
export const color = (r: number, g: number, b: number): color => [r, g, b];

export const mapTuple = (f: (value: number) => number) => (t: tuple) =>
  t.map(f);

export const zipTupleWith = (f: (a: number, b: number) => number) => (
  t1: tuple,
  t2: tuple
) => t1.map((x, i) => f(x, t2[i]));

export const add = zipTupleWith((x, y) => x + y);
export const sub = zipTupleWith((x, y) => x - y);
export const mul = zipTupleWith((x, y) => x * y); // Hadamard product
export const div = zipTupleWith((x, y) => x / y);
export const neg = mapTuple(x => -x);

export const smul = (t: tuple, s: number) => mapTuple(x => x * s)(t);
export const sdiv = (t: tuple, s: number) => mapTuple(x => x / s)(t);

export const magnitude = (t: tuple): number =>
  Math.sqrt(t.reduce((acc, x) => acc + x * x, 0));

export const normalize = (t: tuple): tuple => sdiv(t, magnitude(t));

export const areClose = (
  t1: tuple,
  t2: tuple,
  epsilon: number = 1e-5
): Boolean => magnitude(sub(t1, t2)) < epsilon;

export const dot = (t1: tuple, t2: tuple): number =>
  mul(t1, t2).reduce((acc, x) => acc + x, 0);

// Only valid for 3D vectors
export const cross = (t1: tuple, t2: tuple): tuple =>
  vector(
    t1[1] * t2[2] - t1[2] * t2[1],
    t1[2] * t2[0] - t1[0] * t2[2],
    t1[0] * t2[1] - t1[1] * t2[0]
  );
