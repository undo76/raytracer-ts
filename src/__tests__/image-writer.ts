import Canvas from '../canvas';
import { color } from '../tuple';

describe('Image writer', () => {  
  test('canvas_to_ppm', () => {
    const red = color(2, 0.1, -1);
    const canvas = new Canvas(4, 3, red);
    const ppm = canvas.toPpmString();
    expect(ppm).toMatchSnapshot();
  });
});
