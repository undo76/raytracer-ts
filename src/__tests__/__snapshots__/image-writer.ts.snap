// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image writer canvas_to_ppm 1`] = `
"P3
4 3
255
255 25 0 255 25 0 255 25 0 255 25 0 255 25 0 255 25 0 255 25 0 255 25 0 255 25 0 255 25 0
255 25 0 255 25 0
"
`;
